check for overflow in calculate_nbytes

JacobSzwejbka opened the pull request:
Summary: Got a fuzzer error around overflow here

Differential Revision: D75544079

lucylq commented on runtime/executor/method_meta.cpp:
```diff
@@ -55,12 +55,28 @@ Result<Tag> get_tag(
     n *= sizes[i];
+    // Check for overflow
+    ET_CHECK_MSG(
+      sizes[i] == 0 || n / sizes[i] == prev_n,
```
nit: instead of storing a var for prev_n, wdyt about:
```
ET_CHECK_MSG(sizes[i] == 0 || ((n*sizes[i])/sizes[i]) == n))
```

lucylq commented on runtime/executor/method_meta.cpp:
```diff
@@ -55,12 +55,28 @@ Result<Tag> get_tag(
+  prev_n = n;
+  n = n * elem_size;
+
+  // Check for overflow before multiplying by element size
```
Update the comment, or do it before line 72?

Could do the same thing, 
```
ET_CHECK_MSG(elem_size == 0 || (((n*elem_size)/elem_size) == n);

n = n * elem_size;
```

lucylq reviewed (approved):
Thanks for adding this! Could you also make the same changes to tensor_layout?

https://github.com/pytorch/executorch/blob/14c3b310ee873dbe9f8decf37bf32096b4abcf81/runtime/core/tensor_layout.cpp#L19

JacobSzwejbka committed:
check for overflow in calculate_nbytes (#11217)

Summary:
Got a fuzzer error around overflow here


Test Plan:
Imported from GitHub, without a `Test Plan:` line.

Rollback Plan:

Reviewed By: lucylq

Differential Revision: D75544079

Pulled By: JacobSzwejbka
